### ISSUE TITLE
Linear Attention Mechanism

### ISSUE THREAD

**Issue by parmarsuraj99** [tensorflow/addons#2150]
**Describe the feature and the current behavior/state.**

Are we going to add `LinearAttention`? If yes, I can start working on it

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): [Transformers are RNNs: Fast Autoregressive Transformers with Linear Attention](https://arxiv.org/abs/2006.16236)
- Is there already an implementation in another framework? (if so, where): [Official PyTorch implementation](https://github.com/idiap/fast-transformers/blob/master/fast_transformers/attention/linear_attention.py)
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)** layer

**Who will benefit with this feature?** Building Transformer blocks that are faster with O(N) complexity compared to standard Softmax dot product attnetion

**Any other info.**
[Paper's website](https://linear-transformers.com/)

**Comment by AakashKumarNain** [comment]
Please feel free to open a PR @parmarsuraj99 Thank you

**Comment by saberkun** [comment]
Looking at the pytorch implementation, what's the difference with https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/dense_attention.py?

**Comment by parmarsuraj99** [comment]
In the PyTorch implementation, the authors have implemented many variants of attentions [https://github.com/idiap/fast-transformers/tree/master/fast_transformers/attention](https://github.com/idiap/fast-transformers/tree/master/fast_transformers/attention)

The one referenced above is this specific one
[https://github.com/idiap/fast-transformers/blob/master/fast_transformers/attention/linear_attention.py](https://github.com/idiap/fast-transformers/blob/master/fast_transformers/attention/linear_attention.py)

Major difference is the calculation of Values. Instead of Softmax, they introduced a kernel (here `elu`) on Q and K. Which according to them has shown some improvements.

Major change is focused only in this part.
https://github.com/tensorflow/addons/blob/d466cb8ec3cb8de325b9cb7f48a773f5bbeec808/tensorflow_addons/layers/multihead_attention.py#L223-L239

Can we implement something like callable Attention calculation after calculating Linear projections of inputs?
 something like,
```
query = tf.einsum("...NI , HIO -> ...NHO", query, self.query_kernel)
key = tf.einsum("...MI , HIO -> ...MHO", key, self.key_kernel)
value = tf.einsum("...MI , HIO -> ...MHO", value, self.value_kernel)

output, attn_coef = scaled_dot_product_attention(query, key, value, mask)

or

output, attn_coef = linearized_attention(query, key, value, mask)
```

**Comment by tanzhenyu** [comment]
> In the PyTorch implementation, the authors have implemented many variants of attentions https://github.com/idiap/fast-transformers/tree/master/fast_transformers/attention
> 
> The one referenced above is this specific one
> https://github.com/idiap/fast-transformers/blob/master/fast_transformers/attention/linear_attention.py
> 
> Major difference is the calculation of Values. Instead of Softmax, they introduced a kernel (here `elu`) on Q and K. Which according to them has shown some improvements.
> 
> Major change is focused only in this part.
> https://github.com/tensorflow/addons/blob/d466cb8ec3cb8de325b9cb7f48a773f5bbeec808/tensorflow_addons/layers/multihead_attention.py#L223-L239
> 
> Can we implement something like callable Attention calculation after calculating Linear projections of inputs?
> something like,
> 
> ```
> query = tf.einsum("...NI , HIO -> ...NHO", query, self.query_kernel)
> key = tf.einsum("...MI , HIO -> ...MHO", key, self.key_kernel)
> value = tf.einsum("...MI , HIO -> ...MHO", value, self.value_kernel)
> 
> output, attn_coef = scaled_dot_product_attention(query, key, value, mask)
> 
> or
> 
> output, attn_coef = linearized_attention(query, key, value, mask)
> ```

This seems to be something we could ask user to subclass?

**Comment by saberkun** [comment]
@parmarsuraj99 Thanks!

A bit more concrete idea of subclassing.
There are many innovations about how attention is computed.
Thus, we are trying to have the keras MultiHeadAttention layer being subclassed:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/multi_head_attention.py#L399
A few variants of attention can be accomplished by overriding the compute_attention method.
The variants subclass layers could be good fit to hosted inside packages like addons.

**Comment by parmarsuraj99** [comment]
> 
> 
> @parmarsuraj99 Thanks!
> 
> A bit more concrete idea of subclassing.
> There are many innovations about how attention is computed.
> Thus, we are trying to have the keras MultiHeadAttention layer being subclassed:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/multi_head_attention.py#L399
> A few variants of attention can be accomplished by overriding the compute_attention method.
> The variants subclass layers could be good fit to hosted inside packages like addons.

Thanks

This is exactly what I was referring to implement. 
Subclassing MultiHeadAttention and flexible attention computation.

**Comment by seanpmorgan** [comment]
Given the feedback from TF team, if you would like to submit and maintain a PR for subclassing the Keras MHA that would be okay to proceed.

**Comment by abhishek-niranjan** [comment]
@seanpmorgan  I'd like to contribute to this too.  @parmarsuraj99 let me know if you'd like to work on it together?

**Comment by parmarsuraj99** [comment]
@abhishek-niranjan Sure. I'd really love to collaborate

**Comment by claverru** [comment]
How is this going?

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
